sparse/_common.py

```diff
@@ -25,11 +28,1137 @@
+            for k in a_indices[a_indptr[j] : a_indptr[j + 1]]:
+                if b[j, i] != 0 and mask[k] != i:
+                    mask[k] = i
+                    col_nnz += 1
+        nnz += col_nnz
+        indptr[i + 1] = nnz
+    return nnz
+
+
+@_memoize_dtype
+def _dot_csr_csr_type(dt1, dt2):
+    dtr = np.result_type(dt1, dt2)
+
+    @numba.jit(
+        nopython=True,
+        nogil=True,
+        locals={"data_curr": numba.np.numpy_support.from_dtype(dtr)},
+    )
+    def _dot_csr_csr(
+        out_shape, a_data, b_data, a_indices, b_indices, a_indptr, b_indptr
+    ):  # pragma: no cover
+        """
+        Utility function taking in two ``GCXS`` objects and calculating 
+        their dot product: a @ b for a and b with compressed rows.
+
+        Parameters
+        ----------
+        out_shape : tuple
+            The shape of the output array.
+
+        a_data, a_indices, a_indptr : np.ndarray
+            The data, indices, and index pointer arrays of ``a``.
+
+        b_data, b_indices, b_indptr : np.ndarray
+            The data, indices, and index pointer arrays of ``b``.
+        """
+
+        # much of this is borrowed from:
+        # https://github.com/scipy/scipy/blob/master/scipy/sparse/sparsetools/csr.h
+
+        # calculate nnz before multiplying so we can use static arrays
+        nnz = _csr_csr_count_nnz(out_shape, a_indices, b_indices, a_indptr, b_indptr)
+        n_row, n_col = out_shape
+        indptr = np.empty(n_row + 1, dtype=np.intp)
+        indptr[0] = 0
+        indices = np.empty(nnz, dtype=np.intp)
+        data = np.empty(nnz, dtype=dtr)
+        next_ = np.full(n_col, -1)
+        sums = np.zeros(n_col)
+        nnz = 0
+
+        for i in range(n_row):
+            head = -2
+            length = 0
+            for j, av in zip(
+                a_indices[a_indptr[i] : a_indptr[i + 1]],
+                a_data[a_indptr[i] : a_indptr[i + 1]],
+            ):
+                for k, bv in zip(
+                    b_indices[b_indptr[j] : b_indptr[j + 1]],
+                    b_data[b_indptr[j] : b_indptr[j + 1]],
+                ):
+                    sums[k] += av * bv
+                    if next_[k] == -1:
+                        next_[k] = head
+                        head = k
+                        length += 1
+
+            for _ in range(length):
+                if sums[head] != 0:
+                    indices[nnz] = head
+                    data[nnz] = sums[head]
+                    nnz += 1
+
+                temp = head
+                head = next_[head]
+
+                next_[temp] = -1
+                sums[temp] = 0
+
+            indptr[i + 1] = nnz
+            # ensure sorted indices
+            order = np.argsort(indices[indptr[i] : indptr[i + 1]])
+            data[indptr[i] : indptr[i + 1]] = data[indptr[i] : indptr[i + 1]][order]
+            indices[indptr[i] : indptr[i + 1]] = indices[indptr[i] : indptr[i + 1]][
+                order
+            ]
```

Comment:
Can we possibly sort before putting in the data? I suspect it'll be a lot faster.

Reply:
I'm not quite sure what you mean. Could you give an example?

Reply:
Well, the original CSR-CSR multiplication algorithm doesn't require sorting, correct? I suspect this one only requires it because you're passing in unsorted data in the first place (or iterating over it in a way that's unsorted).

Could you possibly pre-sort the data so the output here will be sorted?

Reply:
The original CSR-CSR multiplication algorithm iterates over the data in a way that's unsorted. In scipy, the resulting csr_matrix has indices that are out-of-order:
```
In [13]: xs.has_sorted_indices
Out[13]: 1
In [14]: res = xs @ ys
In [15]: res.has_sorted_indices
Out[15]: 0
```
Since many of the operations for COO and GCXS rely on maintaining sorted coordinates, I was thinking it would be best to ensure that the result of dot has sorted coords.

Reply:
In that case could you post a benchmark which includes sorting for the SciPy version?

Reply:
```
%%timeit
x @ y # coo coo
10.2 ms ± 51.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%%timeit
res = xs @ ys
res.sort_indices()
4.05 ms ± 8.83 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
I'm going to test doing just one sort at the end to see how that does.

Reply:
```
%%timeit
x @ y # coo coo - one sort at the end
7.55 ms ± 96.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%%timeit
res = xs @ ys
res.sort_indices()
4.04 ms ± 15.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
It looks like doing one sort at the end is better.

Reply:
Could you also post some GCXS-GCXS benchmarks for direct comparison with SciPy?

Reply:
```
%%timeit
x @ y # csr csr
6.32 ms ± 107 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

x = x.change_compressed_axes([1])
y = y.change_compressed_axes([1])
x @ y

%%timeit
x @ y # csc csc
6.11 ms ± 46.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

x = x.tocoo()
y = y.tocoo()
x @ y

%%timeit
x @ y # coo coo
7.23 ms ± 93.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%%timeit
res = xs @ ys
res.sort_indices()
4.08 ms ± 49.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```